core/domain/wipeout_service.py

```diff
@@ -109,11 +107,109 @@ def _delete_user_models(user_id):
+                [base_models.DELETION_POLICY.KEEP,
+                 base_models.DELETION_POLICY.NOT_APPLICABLE]):
             model_class.apply_deletion_policy(user_id)
 
 
+def _hard_delete_explorations_and_collections(pending_deletion_model):
+    """Hard delete the exploration and collection models that are private and
+    solely owned by the user.
+
+    Args:
+        pending_deletion_model: PendingDeletionRequestModel.
+    """
+    exp_services.delete_explorations(
+        pending_deletion_model.id,
+        pending_deletion_model.exploration_ids,
+        force_deletion=True)
+    collection_services.delete_collections(
+        pending_deletion_model.id,
+        pending_deletion_model.collection_ids,
+        force_deletion=True)
+
+
+def _generate_activity_to_user_ids_mapping(activity_ids):
+    """Generate mapping from activity IDs to user IDs.
+
+    Args:
+        activity_ids: list(str). List of activity IDs for which to generate
+            new user IDs.
+
+    Returns:
+        dict(str, str). Mapping between the activity IDs and pseudonymized
+        user IDs.
+    """
+    return {
+        activity_id: user_models.PseudonymizedUserModel.get_new_id()
+        for activity_id in activity_ids
+    }
+
+
+def _delete_story_models(pending_deletion_model):
+    """Pseudonymize the story models for the user with user_id.
+
+    Args:
+        pending_deletion_model: PendingDeletionRequestModel.
+    """
+    user_id = pending_deletion_model.id
+    metadata_models = story_models.StorySnapshotMetadataModel.query(
+        story_models.StorySnapshotMetadataModel.committer_id == user_id
+    ).fetch()
+    story_ids = set([
+        model.get_unversioned_instance_id() for model in metadata_models])
+    commit_log_models = story_models.StoryCommitLogEntryModel.query(
+        story_models.StoryCommitLogEntryModel.user_id == user_id
+    ).fetch()
+    story_ids = story_ids | set(model.story_id for model in commit_log_models)
+
+    if not pending_deletion_model.story_mappings:
+        pending_deletion_model.story_mappings = (
+            _generate_activity_to_user_ids_mapping(story_ids))
+        pending_deletion_model.put()
+
+    def _pseudonymize_models(story_related_models, pseudonymized_user_id):
+        """Pseudonymize user ID fields in the models.
+
+        Args:
+            story_related_models: list(ndb.Model). Models which user IDs should
+                be pseudonymized.
+            pseudonymized_user_id: str. New pseudonymized user ID to be used for
+                the models.
+        """
+        metadata_models = [
+            model for model in story_related_models
+            if isinstance(model, story_models.StorySnapshotMetadataModel)]
+        for metadata_model in metadata_models:
+            metadata_model.committer_id = pseudonymized_user_id
+        story_models.StorySnapshotMetadataModel.put_multi(
+            metadata_models, update_last_updated_time=False)
```

Comment:
This one does seem like an actual update, right? (It's not one that's applied in bulk to all models.)